src/net/sf/mpxj/primavera/UserFieldCounters.java

```diff
@@ -48,36 +52,49 @@ public UserFieldCounters()
     * to a user defined data type.
     *
     * @param type target type
-    * @param fieldName field name overriding the default
+    * @param fieldName field names overriding the default
     */
-   public void setFieldNameForType(UserFieldDataType type, String fieldName)
+   public void setFieldNamesForType(UserFieldDataType type, String[] fieldNames)
    {
-      m_names[type.ordinal()] = fieldName;
+      m_names[type.ordinal()] = fieldNames;
    }
 
    /**
-    * Generate the next name for a user defined field.
+    * Generate the next available field for a user defined field.
     *
+    * @param clazz class of the desired field enum
     * @param type user defined field type.
-    * @return field name
+    * @return field of specified type
     */
-   public String nextName(UserFieldDataType type)
+   public <E extends Enum<E> & FieldType> E nextField(Class<E> clazz, UserFieldDataType type)
    {
-      int counter = ++m_counters[type.ordinal()];
-      return type.getDefaultFieldName() + counter;
+      for (String name : m_names[type.ordinal()])
+      {
+         int i = NumberHelper.getInt(m_counters.get(name)) + 1;
+         try
+         {
+            E e = Enum.valueOf(clazz, name + i);
+            m_counters.put(name, Integer.valueOf(i));
+            return e;
+         }
+         catch (IllegalArgumentException ex)
+         {
+            // try the next name
+         }
+      }
+
+      // no more fields available
+      throw new IllegalArgumentException("No fields for type " + type + " available");
```

Comment:
I think I may have filed that original bug report :) The exception throw here gets caught and swallowed by the try catch in `addUserDefinedField` and that field is just dropped as you mention.

Some of my users had 100+ user defined fields and complained about the 30 field limit; thus why I expanded capacity by filling into the enterprise fields. I couldn't come up with a better way to expand capacity without altering the core MPXJ MS Project model.